sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/TableIdentifierParserSuite.scala

```diff
@@ -285,334 +290,73 @@ class TableIdentifierParserSuite extends SparkFunSuite with SQLHelper {
-    "column",
-    "constraint",
-    "create",
-    "cross",
-    "current_date",
-    "current_time",
-    "current_timestamp",
-    "current_user",
-    "day",
-    "distinct",
-    "else",
-    "end",
-    "escape",
-    "except",
-    "false",
-    "fetch",
-    "for",
-    "foreign",
-    "from",
-    "full",
-    "grant",
-    "group",
-    "having",
-    "hour",
-    "in",
-    "inner",
-    "intersect",
-    "into",
-    "join",
-    "is",
-    "leading",
-    "left",
-    "minute",
-    "month",
-    "natural",
-    "not",
-    "null",
-    "on",
-    "only",
-    "or",
-    "order",
-    "outer",
-    "overlaps",
-    "primary",
-    "references",
-    "right",
-    "select",
-    "semi",
-    "session_user",
-    "minus",
-    "second",
-    "some",
-    "table",
-    "then",
-    "to",
-    "trailing",
-    "union",
-    "unique",
-    "unknown",
-    "user",
-    "using",
-    "when",
-    "where",
-    "with",
-    "year")
+  private def parseAntlrGrammars(startTag: String, endTag: String)
+      (f: PartialFunction[String, Option[String]]): Set[String] = {
+    // We need to map a symbol string to actual literal strings
+    // in case that they have different strings.
+    val symbolsToNeedRemap = Map(
+      "DATABASES" -> Seq("DATABASES", "SCHEMAS"),
+      "RLIKE" -> Seq("RLIKE", "REGEXP"),
+      "SETMINUS" -> Seq("MINUS"),
+      "TEMPORARY" -> Seq("TEMPORARY", "TEMP")
+    )
+    val keywords = new mutable.ArrayBuffer[String]
+    val default = (_: String) => None
+    var start = false
+    for (line <- sqlSyntaxDefs) {
+      if (line.trim.startsWith(startTag)) {
+        start = true
+      } else if (line.trim.startsWith(endTag)) {
+        start = false
+      } else if (start) {
+        f.applyOrElse(line, default).foreach { symbol =>
+          if (symbolsToNeedRemap.contains(symbol)) {
+            keywords ++= symbolsToNeedRemap(symbol)
+          } else {
+            keywords += symbol
+          }
+        }
+      }
+    }
+    keywords.map(_.trim.toLowerCase(Locale.ROOT)).toSet
+  }
+
+  private def parseSyntax(startTag: String, endTag: String): Set[String] = {
+    val kwDef = """\s*[\|:]\s*([A-Z_]+)\s*""".r
+    parseAntlrGrammars(startTag, endTag) {
+      // Parses a pattern, e.g., `    | AFTER`
+      case kwDef(symbol) => Some(symbol)
+    }
+  }
+
+  // All the SQL keywords defined in `SqlBase.g4`
+  val allCandidateKeywords = {
+    val kwDef = """([A-Z_]+):.+;""".r
+    val keywords = parseAntlrGrammars(
+        "//--SPARK-KEYWORD-LIST-START", "//--SPARK-KEYWORD-LIST-END") {
+      // Parses a pattern, e.g., `AFTER: 'AFTER';`
+      case kwDef(symbol) => Some(symbol)
+    }
+    keywords
+  }
+
+  val nonReservedKeywordsInAnsiMode = parseSyntax(
+    "//--ANSI-NON-RESERVED-START", "//--ANSI-NON-RESERVED-END")
 
-  val nonReservedKeywordsInAnsiMode = allCandidateKeywords -- reservedKeywordsInAnsiMode
+  val reservedKeywordsInAnsiMode = allCandidateKeywords -- nonReservedKeywordsInAnsiMode
```

Comment:
If we remove the tags, will the related tests fail? I have this question because these tests seems do `foreach` on these lists. If we remove the tags and get an empty list, seems the tests won't fail too?

Reply:
Ah, nice suggestion. In that case, the tests should fail. I'll check and update.

Reply:
Could you check the latest commit? Is that okay to you?

Reply:
Yeah, thanks. Looks good.